nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -498,6 +507,10 @@ in
 
       boot = {
         kernelModules = [ "zfs" ];
+        # https://github.com/openzfs/zfs/issues/260
+        # https://github.com/openzfs/zfs/issues/12842
+        # https://github.com/NixOS/nixpkgs/issues/106093
+        kernelParams = lib.optionals (!config.boot.zfs.allowHibernation) [ "nohibernate" ];
```

Comment:
I don't think this is the right workaround. Because it allows people to assign a `swapDevices` but that not taking any effect. I think it would be better to have an assertion like this:

```nix
{
  assertions = [{
    assertion = (config.boot.zfs.enable && config.swapDevices != []) -> config.config.boot.zfs.allowHibernation;
    message = "Using swap with ZFS can corrupt data. Either disable swap by removing all `swapDevices` entries, or if you know what you're doing, ignore this assertion by enabling `boot.zfs.allowHibernation`";
  }];
```

Reply:
We should also have another assertion which throws when ZFS is enabled and you have a swap file on a ZFS filesystem

Reply:
> Using swap with ZFS can corrupt data.

Does this apply even for a swap partition that's not on a ZVOL or otherwise on ZFS?

Reply:
> I don't think this is the right workaround. Because it allows people to assign a `swapDevices` but that not taking any effect. I think it would be better to have an assertion like this:
> 
> ```nix
> {
>   assertions = [{
>     assertion = (config.boot.zfs.enable && config.swapDevices != []) -> config.config.boot.zfs.allowHibernation;
>     message = "Using swap with ZFS can corrupt data. Either disable swap by removing all `swapDevices` entries, or if you know what you're doing, ignore this assertion by enabling `boot.zfs.allowHibernation`";
>   }];
> ```

I just got bit by this. I'm only using ZFS on a block file in an EXT4 filesystem so I'm not concerned with hibernating ZFS filesystems. I think this suggestion would aid discoverability.

Reply:
I think there's a misunderstanding in @infinisil's year old comment. The issue is not that you cannot have swap devices while using ZFS; you totally can (though, for *other* unrelated reasons, those cannot be stored on ZFS). The issue is that you cannot hibernate while ZFS pools are imported, period, no matter where the swap is stored, or else those pools have a chance of being irreparably corrupted.

Point being: ZFS + swap on non-ZFS: Good!. ZFS + hibernation in any case whatsoever: Bad! Hence what we have now is the right fix. If you want to ignore it, well it is configurable.

Reply:
> The issue is that you cannot hibernate while ZFS pools are imported, period, no matter where the swap is stored, or else those pools have a chance of being irreparably corrupted.

Makes sense! Though, the discoverability for [hibernation being disabled](https://github.com/NixOS/nixpkgs/blob/d2e4de209881b38392933fabf303cde3454b0b4c/nixos/modules/tasks/filesystems/zfs.nix#L545) is not great. I only realized ZFS was the problem after reading [this](https://discourse.nixos.org/t/hibernate-doesnt-work-anymore/24673/12?u=ryang) discourse post. I'm really not using ZFS in anger, so I'm happy to risk data loss for being able to hibernate my laptop, and I think @infinisil's suggestion would make this clear to new users.

Reply:
infinisil's suggestion would disable swap entirely with ZFS, which is certainly *not* desirable. We don't really have a way to know at eval-time whether a user intends to use hibernation, so it's difficult to warn about.

Actually, we *should* probably have an assertion that `boot.resumeDevice` cannot be set while `allowHibernation` is false, and that would cover *some* use cases. But NixOS has auto-detection of resume devices in stage 1 (if you're using EFI and scripted initrd exactly because... reasons), so this is very often not set.

Reply:
> infinisil's suggestion would disable swap entirely with ZFS, which is certainly not desirable. We don't really have a way to know at eval-time whether a user intends to use hibernation, so it's difficult to warn about.

Hmm, how about:
```
{
  assertions = [{
    assertion = (config.boot.zfs.enable && ! (builtins.elem "nohibernate" config.boot.kernelParams)) -> config.config.boot.zfs.allowHibernation;
    message = "Using hibernation with ZFS can corrupt data. Either disable hibernation with `boot.kernelParams = [ "nohibernate" ]` or if you know what you're doing, ignore this assertion by enabling `boot.zfs.allowHibernation`";
  }];
```

Reply:
As also mentioned in https://github.com/NixOS/nixpkgs/pull/171680#issuecomment-1301009811, I've been using ZFS with swap (on a separate partition) with hibernation for a while now and nothing has corrupted :shrug:. I am using a patch on top of Nixpkgs though: https://github.com/infinisil/nixpkgs/commit/448fe5db58cddfcc9258a71ca0c4c11d19ca1ecb. This makes swap restore before ZFS imports.

Reply:
@infinisil that patch *helps*. Certainly, it *should* be quite rare with that patch. But "quite a rare, but known, source of data loss" is not something we want people using by accident.

@RyanGibb That assertion doesn't really *do* anything though? `allowHibernation` *implies* that `nohibernate` is not in the cmdline. Unless the user manually added it themselves *and* enabled `allowHibernation`. While that would be a nonsensical thing for them to do, it would at least be safe, and nothing to worry about, and means that they explicitly added both configuration lines.